README
=========================================================================
==========================   CAMPCASTER   ===============================
=========================================================================
Campcaster is the first free and open radio management software that 
provides live studio broadcast capabilities as well as remote automation
in one integrated system.

Campcaster has three audio outputs: one for live broadcasting, one for
previewing audio, and one for automated scheduled playback.

Major features of Campcaster:
 * Live, in-studio playout.  Campcaster Studio enables station personnel
   to play out both individual sound files and playlists such as 
   advertising blocks or entire programs.
 * Web-based remote station management.  Authorized personnel can add
   program material, create playlists, and schedule programming all via 
   Campcaster's web interface.
 * Automation.  Campcaster has a scheduler function that enables users to
   set playlists for playback at a date and time of their choosing.
   Playlists can be played back multiple times.
 * Playlists.  Campcaster allows users to create playlists that can be 
   either played on-demand in the Studio or in automated mode.  Playlists
   can be nested inside each other, so for example, a playlist with a 
   block of ads can be inside of a playlist with a music program.
 * Centralized archives of station program material.  Station personnel
   have a searchable, browseable central archive that is accessible both
   in the office and via the web.
 * Solid, fast playback.  Campcaster uses the popular Gstreamer 
   multimedia framework for clean, reliable, fast playback. 
 * Program sharing.  Stations can share both playlists and files with a 
   designated, centralized 'network hub.'  This enables the creation of 
   'virtual radio networks' connected via the Internet. 
 * Search-based backup.  You can backup all station archives at the click
   of a button, or only backup certain files based on search terms, so 
   that backup files can be made of only a certain program or of all 
   files created by a certain user.
 * Localization.  Campcaster is multilingual and supports Unicode.  It
   also is very easy to create language localizations.
 * No restrictions.  Multiple instances of the Campcaster Studio program 
   can access the same archive on the office network, and there are no 
   worries about hardware dongles or keys limiting access on multiple PCs.
 * Innovative design.  Campcaster's user interface – the product of 
   Parsons School of Design's “Design and Communication” program -  
   introduces a number of innovations, including the 'Scratchpad', 
   a list of recently-used files and playlists.
 * Open, extensible architecture.  Stations are free to extend and alter 
   all parts of the Campcaster program code, which makes extensive use 
   of XML-RPC APIs.


INSTALLATION
------------
If you are running Ubuntu 10.04 (Lucid), the easiest way to install is to add:

  deb http://apt.64studio.com/backports/ lucid-backports main

to your sources.list file, and then type:

  sudo apt-get update
  sudo apt-get install campcaster-studio campcaster-station campcaster-libs

 * Building & Installing on all versions of Ubuntu/Kubuntu etc.
   doc/installUbuntu.html

 * Installing on other platforms
   doc/install.html

 * Developing on all versions of GNU/Linux
   http://wiki.sourcefabric.org/display/CC/Campcaster+Dev+Site

 * Developing on other platforms
   doc/gettingStarted.html

Documentation
-------------
 * User manual
   http://en.flossmanuals.net/Campcaster/


-------------------------------------------------------------------------

    Copyright (c) 2010 Sourcefabric

    This file is part of the Campcaster project.
    http://campcaster.sourcefabric.org/
    To report bugs, send an e-mail to contact@sourcefabric.org

    Campcaster is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Campcaster is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Campcaster; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

-------------------------------------------------------------------------



debian/changelog
campcaster (1.4.1-3) 64studio; urgency=low

  * Revert to default database password to preserve compatibility

 -- Daniel James <daniel@64studio.com>  Wed, 24 Nov 2010 11:52:21 +0000

campcaster (1.4.1-2) 64studio; urgency=low

  * Removed LD_LIBRARY_PATHS from makefiles

 -- Daniel James <daniel@64studio.com>  Tue, 23 Nov 2010 14:28:40 +0000

campcaster (1.4.1-1) 64studio; urgency=low

  * Update my email address.
  * Fix small issues in the manpages.
  * Bump Standards.
  * Update Homepage field.

 -- Alessio Treglia <alessio@debian.org>  Sat, 20 Nov 2010 11:58:21 +0100

campcaster (1.4.1-0) 64studio; urgency=low

  * ongoing 1.4.x support

 -- Robin Gareus <robin@gareus.org>  Wed, 17 Nov 2010 21:59:46 +0100

campcaster (1.4.0+r3436-1) 64studio; urgency=low

  * Fix for crash on 64-bit systems in GLiveSupport.cxx

 -- Daniel James <daniel@64studio.com>  Mon, 04 Oct 2010 16:17:29 +0100

campcaster (1.4.0+r3436-0) 64studio; urgency=low

  * rebased to new upstream version

 -- Robin Gareus <robin@gareus.org>  Fri, 16 Apr 2010 11:59:57 +0200

campcaster (1.4.0+r3427) 64studio; urgency=low

  * Initial release (Closes: #568530) (LP: #250397)

 -- Robin Gareus <robin@gareus.org>  Fri, 05 Feb 2010 16:35:44 +0100


campcaster/src/modules/core/src/Md5.cxx
/*------------------------------------------------------------------------------

    Copyright (c) 2004 Media Development Loan Fund
 
    This file is part of the Campcaster project.
    http://campcaster.campware.org/
    To report bugs, send an e-mail to bugs@campware.org
 
    Campcaster is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
  
    Campcaster is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with Campcaster; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

  ------------------------------------------------------------------------------
 
    This class is based on the following with minor modifications
    (see http://userpages.umbc.edu/~mabzug1/cs/md5/md5.html).

  ------------------------------------------------------------------------------

// Md5.CC - source code for the C++/object oriented translation and 
//          modification of Md5.

// Translation and modification (c) 1995 by Mordechai T. Abzug 

// This translation/ modification is provided "as is," without express or 
// implied warranty of any kind.

// The translator/ modifier does not claim (1) that Md5 will do what you think 
// it does; (2) that this translation/ modification is accurate; or (3) that 
// this software is "merchantible."  (Language for this disclaimer partially 
// copied from the disclaimer below).

  ------------------------------------------------------------------------------

    Which was based on:

  ------------------------------------------------------------------------------

   Md5.H - header file for Md5C.C
   MDDRIVER.C - test driver for MD2, MD4 and Md5

   Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. Md5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. Md5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.

  ------------------------------------------------------------------------------

    Author   : $Author: fgerlits $
    Version  : $Revision: 2329 $
    Location : $URL: svn://code.campware.org/campcaster/trunk/campcaster/src/modules/core/src/Md5.cxx $

------------------------------------------------------------------------------*/

/* ============================================================ include files */

#include "LiveSupport/Core/Md5.h"

#include <assert.h>
#include <string>
#include <iostream>


using namespace LiveSupport::Core;

/* ===================================================  local data structures */


/* ================================================  local constants & macros */


/* ===============================================  local function prototypes */


/* =============================================================  module code */

/*------------------------------------------------------------------------------
 *  Construct from a string.
 *----------------------------------------------------------------------------*/
Md5::Md5(const std::string &s)                  throw(std::invalid_argument)
{
  init();
  update (s);
  finalize();
  calcNumericRepresentation();
}


/*------------------------------------------------------------------------------
 *  Construct from an istream.
 *----------------------------------------------------------------------------*/
Md5::Md5(std::istream& stream)                  throw(std::invalid_argument)
{
  init();
  update (stream);
  finalize();
  calcNumericRepresentation();
}


/*------------------------------------------------------------------------------
 *  Construct from a file.
 *----------------------------------------------------------------------------*/
Md5::Md5(FILE *file)                            throw(std::invalid_argument)
{
  init();
  update(file);
  finalize ();
  calcNumericRepresentation();
}


/*------------------------------------------------------------------------------
 *  Output the md5 sum as a hexadecimal string.
 *----------------------------------------------------------------------------*/
std::string Md5::hexDigest()                    throw()
{
    std::stringstream strStr;
    
    for (int i=0; i<16; i++) {
        strStr << std::hex << std::setw(2) << std::setfill('0') 
               << int(digest[i]);
    }
    
    return strStr.str();
}


/*------------------------------------------------------------------------------
 *  Calculate the numeric representation of the checksum
 *----------------------------------------------------------------------------*/
void
Md5 :: calcNumericRepresentation(void)                  throw ()
{
    low64  = (uint64_t) digest[15];
    low64 |= ((uint64_t) digest[14]) << 8;
    low64 |= ((uint64_t) digest[13]) << 16;
    low64 |= ((uint64_t) digest[12]) << 24;
    low64 |= ((uint64_t) digest[11]) << 32;
    low64 |= ((uint64_t) digest[10]) << 40;
    low64 |= ((uint64_t) digest[9])  << 48;
    low64 |= ((uint64_t) digest[8])  << 56;

    high64  = (uint64_t) digest[7];
    high64 |= ((uint64_t) digest[6]) << 8;
    high64 |= ((uint64_t) digest[5]) << 16;
    high64 |= ((uint64_t) digest[4]) << 24;
    high64 |= ((uint64_t) digest[3]) << 32;
    high64 |= ((uint64_t) digest[2]) << 40;
    high64 |= ((uint64_t) digest[1]) << 48;
    high64 |= ((uint64_t) digest[0]) << 56;
}


/*------------------------------------------------------------------------------
 *  Convert the md5 sum to a hexadecimal string.
 *----------------------------------------------------------------------------*/
Md5::operator std::string()                     throw()
{
    return hexDigest();
}


/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ private methods */

// Md5 block update operation. Continues an Md5 message-digest
// operation, processing another message block, and updating the
// context.

void Md5::update (uint1 *input, uint4 input_length)
{
    uint4 input_index, buffer_index;
    uint4 buffer_space;                // how much space is left in buffer

    if (finalized){  // so we can't update!
        throw std::invalid_argument("Md5::update: "
                                    "Can't update a finalized digest!");
    }

    // Compute number of bytes mod 64
    buffer_index = (unsigned int)((count[0] >> 3) & 0x3F);

    // Update number of bits
    if (  (count[0] += ((uint4) input_length << 3))<((uint4) input_length << 3) )
        count[1]++;

    count[1] += ((uint4)input_length >> 29);


    buffer_space = 64 - buffer_index;  // how much space is left in buffer

    // Transform as many times as possible.
    if (input_length >= buffer_space) { // ie. we have enough to fill the buffer
        // fill the rest of the buffer and transform
        memcpy (buffer + buffer_index, input, buffer_space);
        transform (buffer);

        // now, transform each 64-byte piece of the input, bypassing the buffer
        for (input_index = buffer_space; input_index + 63 < input_length; 
	                                                       input_index += 64) {
            transform (input+input_index);
        }
    buffer_index = 0;  // so we can buffer remaining
    } else {
        input_index=0;     // so we can buffer the whole input
    }

    // and here we do the buffering:
    memcpy(buffer+buffer_index, input+input_index, input_length-input_index);
}



// Md5 update for files.
// Like above, except that it works on files (and uses above as a primitive.)

void Md5::update(FILE *file)
{
    uint1 buffer[1024];
    uint4 len;

    while ((len=fread(buffer, 1, 1024, file)))
        update(buffer, len);

//    deleted by me (fgerlits) to make it work more like the istream version
//    fclose (file);
}






// Md5 update for istreams.
// Like update for files; see above.

void Md5::update(std::istream& stream)
{
    uint1 buffer[1024];
    uint4 len;

    while (stream.good()){   // note that return value of read is unusable.
        stream.read((char *)buffer, 1024);
        len=stream.gcount();
        update(buffer, len);
    }
}






// Md5 update for strings.
// Like update for files; see above.

void Md5::update(const std::string &s)
{
    uint1 buffer[1024];
    uint4 len;
    uint4 pos = 0;

    while (pos < s.size()){
        len = s.copy((char *)buffer, 1024, pos);
        update(buffer, len);
        pos += len;
    }
}






// Md5 finalization. Ends an Md5 message-digest operation, writing the
// the message digest and zeroizing the context.


void Md5::finalize ()       throw (std::invalid_argument)
{
    uint1 bits[8];
    uint4 index, padLen;
    static uint1 PADDING[64]={
        0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };
    
    if (finalized){
        throw std::invalid_argument("Md5::finalize: "
                                    "Already finalized this digest!");
    }
    
    // Save number of bits
    encode (bits, count, 8);
    
    // Pad out to 56 mod 64.
    index = (uint4) ((count[0] >> 3) & 0x3f);
    padLen = (index < 56) ? (56 - index) : (120 - index);
    update (PADDING, padLen);
    
    // Append length (before padding)
    update (bits, 8);
    
    // Store state in digest
    encode (digest, state, 16);
    
    // Zeroize sensitive information
    memset (buffer, 0, sizeof(*buffer));
    
    finalized=1;
}




// PRIVATE METHODS:



void Md5::init(){
  finalized=0;  // we just started!

  // Nothing counted, so count=0
  count[0] = 0;
  count[1] = 0;

  // Load magic initialization constants.
  state[0] = 0x67452301;
  state[1] = 0xefcdab89;
  state[2] = 0x98badcfe;
  state[3] = 0x10325476;
}



// Constants for Md5Transform routine.
// Although we could use C++ style constants, defines are actually better,
// since they let us easily evade scope clashes.

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21




// Md5 basic transformation. Transforms state based on block.
void Md5::transform (uint1 block[64]){

  uint4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  decode (x, block, 64);

  assert(!finalized);  // not just a user error, since the method is private

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  // Zeroize sensitive information.
  memset ( (uint1 *) x, 0, sizeof(x));

}



// Encodes input (UINT4) into output (unsigned char). Assumes len is
// a multiple of 4.
void Md5::encode (uint1 *output, uint4 *input, uint4 len) {

  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
    output[j]   = (uint1)  (input[i] & 0xff);
    output[j+1] = (uint1) ((input[i] >> 8) & 0xff);
    output[j+2] = (uint1) ((input[i] >> 16) & 0xff);
    output[j+3] = (uint1) ((input[i] >> 24) & 0xff);
  }
}




// Decodes input (unsigned char) into output (UINT4). Assumes len is
// a multiple of 4.
void Md5::decode (uint4 *output, uint1 *input, uint4 len){

  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
    output[i] = ((uint4)input[j]) | (((uint4)input[j+1]) << 8) |
      (((uint4)input[j+2]) << 16) | (((uint4)input[j+3]) << 24);
}





// Note: Replace "for loop" with standard memcpy if possible.
void Md5::memcpy (uint1 *output, uint1 *input, uint4 len){

  unsigned int i;

  for (i = 0; i < len; i++)
    output[i] = input[i];
}



// Note: Replace "for loop" with standard memset if possible.
void Md5::memset (uint1 *output, uint1 value, uint4 len){

  unsigned int i;

  for (i = 0; i < len; i++)
    output[i] = value;
}



// ROTATE_LEFT rotates x left n bits.

inline unsigned int Md5::rotate_left  (uint4 x, uint4 n){
  return (x << n) | (x >> (32-n))  ;
}




// F, G, H and I are basic Md5 functions.

inline unsigned int Md5::F            (uint4 x, uint4 y, uint4 z){
  return (x & y) | (~x & z);
}

inline unsigned int Md5::G            (uint4 x, uint4 y, uint4 z){
  return (x & z) | (y & ~z);
}

inline unsigned int Md5::H            (uint4 x, uint4 y, uint4 z){
  return x ^ y ^ z;
}

inline unsigned int Md5::I            (uint4 x, uint4 y, uint4 z){
  return y ^ (x | ~z);
}



// FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
// Rotation is separate from addition to prevent recomputation.


inline void Md5::FF(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
		    uint4  s, uint4 ac){
 a += F(b, c, d) + x + ac;
 a = rotate_left (a, s) +b;
}

inline void Md5::GG(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
		    uint4 s, uint4 ac){
 a += G(b, c, d) + x + ac;
 a = rotate_left (a, s) +b;
}

inline void Md5::HH(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
		    uint4 s, uint4 ac){
 a += H(b, c, d) + x + ac;
 a = rotate_left (a, s) +b;
}

inline void Md5::II(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
			     uint4 s, uint4 ac){
 a += I(b, c, d) + x + ac;
 a = rotate_left (a, s) +b;
}




